Introduce Telemetry Schemas (#152)
open-telemetry/oteps: text/0152-telemetry-schemas.md

implementations will no longer work correctly together.

Essentially there is a coupling between 3 parties: 1) OpenTelemetry semantic
conventions, 2) telemetry sources and 3) telemetry consumers. The coupling
complicates the independent evolution of these 3 parties.

We recognize the following needs:

- OpenTelemetry semantic conventions need to evolve over time. When conventions
  are first defined, mistakes are possible and we may want to fix the mistakes
  over time. We may also want to change conventions to re-group the attributes
  into different namespaces as our understanding of the attribute taxonomy
  improves.

- Telemetry sources over time may want to change the schema of the telemetry
  they emit. This may be because for example the semantic conventions evolved
  and we want to make our telemetry match the newly introduced conventions.

- In an observability system there may simultaneously exist telemetry sources
  that produce data that confirms to different telemetry schemas because
  different sources evolve at a different pace and are implemented and
  controlled by different entities.

- Telemetry consumers have a need to understand what schema a particular piece

>   of received telemetry confirms to. The consumers also need a way to be able to

Reviewer: ```suggestion
  of received telemetry conforms to. The consumers also need a way to be able to
```

Author: Fixed.

  that reason this document is concerned with _changes_ to the schema as opposed
  to the _full state_ of the schema. We do not preclude this though: the schema
  file format is extensible and in the future may allow defining the full state
  of the schema, see [Current State in Schema](#current-state-in-schema) in
  Future Possibilities section).

- We intentionally limit the types of transformations of schemas to the bare
  minimum that is necessary to handle the most common changes that we believe
  OpenTelemetry Semantic Conventions will require in the near future. More types
  of transformations [may be proposed](#other-transformation-types) in the
  future. This proposal does not attempt to support a comprehensive set of
  possible transformation types that can handle all possible changes to schemas
  that we can imagine. That would be too complicated and very likely
  superfluous. Any new transformation types should be proposed and added in the
  future to the schema file format when there is an evidence that they are
  necessary for the evolution of OpenTelemetry.

## Use Cases

This section shows a couple interesting use-cases for the telemetry schemas

> (other uses-cases are also possible, this is not an exhaustive list).

Reviewer 2: Regarding the use case where the backend is not aware of schemas and the collector is not used: should there be any SDK functionality for this use case, or is it left to each individual exporter to implement schema conversions?

Author: It is a possibility. We will need to decide if we want to put that burden on each language SDK vs implementing schema conversion once in the Collector.

Reviewer 2: It would indeed be a burden on languages SDKS, however, I think when not providing this feature on SDK level we end up with inconsistent and unexpected user experiences.

E. g., the Jaeger and Zipkin exporters which every SDK ships by default wouldn't provide any schema mapping capabilities.

Apart from that I really like this proposal and the direction it's heading towards.

Reviewer 3: I would vote against re-implementing schema conversion in every language.

Reviewer 4: That would require everyone to use the collector?

Author: >That would require everyone to use the collector?

Either use a Collector or use a backend which understands schemas.

Alternatively, if there is no Collector and the backend does not understand schemas the system will operate just as it does today, without being aware of schemas. That is not a catastrophic situation, only things that rely on the schema may be broken, the rest should work. So the degraded behaviour in systems that do not understand schemas is no worse than we have today.

the schema (e.g. the latest OpenTelemetry schema file can be packaged at build
time with OpenTelemetry Collector's schema translation processor).

## Schema Version Number

Version number follows the MAJOR.MINOR.PATCH format, similar to semver 2.0. Note
that we currently only use a subset of semver functionality, see below.

As the schema evolves over time the version number SHOULD change according to
the following rules:

- MAJOR number must increase when the new version is incompatible with the
  previous version.

- MINOR number must increase when the new version is compatible with the
  previous version.

- PATCH number is reserved and SHOULD be equal to 0.

"Compatible" in this context means that the schema file defines transformations
that can be applied to convert telemetry data from the previous version to the
new version and vice versa (note: the conversions may not always be possible in
both directions, see later in the text).

"Incompatible" means that transformations between the new version and the
previous version are not possible.

Note: version numbers use the
[ordering rules](https://semver.org/#spec-item-11) defined by semver 2.0
specification.

## Schema File

> Schema File is a YAML file that describes the schema of a particular version. It

Reviewer: ```suggestion
A Schema File is a YAML file that describes the schema of a particular version. It
```

Author: Fixed.

  // "instrumentation_library_logs" except the logs which have a schema_url
  // specified in the nested InstrumentationLibraryLogs message.
  string schema_url = 3;
}
message InstrumentationLibraryLogs {
  ...
  // This schema_url applies to all logs in the "logs" field regardless of the
  // value of the schema_url field in the outer ResourceLogs message.
  string schema_url = 3;
}
```

The schema_url field in the ResourceSpans, ResourceMetrics, ResourceLogs
messages applies to the contained Resource, Span, SpanEvent, Metric, LogRecord
messages.

The schema_url field in the InstrumentationLibrarySpans message applies to the
contained Span and SpanEvent messages.

The schema_url field in the InstrumentationLibraryMetrics message applies to the
contained Metric messages.

The schema_url field in the InstrumentationLibraryLogs message applies to the
contained LogRecord messages.

If schema_url field is non-empty both in Resource* message and in the contained
InstrumentationLibrary* message then the value in InstrumentationLibrary*
message takes the precedence.

## API and SDK Changes

OpenTelemetry API and SDK require the following changes:

> - Add method SetSchema(schema_url) to Tracer. After this call all telemetry

Reviewer 5: I think it would be better to only allow this to be called once, e.g. by making it an optional argument to the Provider.

Author: This is good point and I was thinking about it, but I was not sure we want to alter the signature of an API that is already declared stable (GetTracer). If I am not wrong it would be a backward compatible API change but it will likely be a breaking ABI change (I am forgetting now if we require ABI compatibility for languages like C++).

Alternatively for languages which allow overloading we could introduce a new GetMeter/GetTracer/GetLogEmitter overload which takes 3 parameters (the 3rd parameter being schema_url).

I am open for suggestions.

Reviewer 5: We need to have a design principle that certain API methods are extensible, and use language-specific constructs to achieve that. E.g. in Java the tracer shouldn't be accessed via `provider.getTracer`, but using a builder. In Python it can be getTracer method with kw args.

Reviewer: I agree that this should be set only once, preferably at creation time, and once the tracer has been used at all, that it should not be able to have it's schema value changed. 

For instance, I think in Java, we'd probably just add an overload to the TracerProvider.get() method to allow the specification of a schema URL, and under the hood, we'd default it to whatever schema version the semantic convention library aligned with the SDK version used (or something functionally equivalent to this... we might choose to introduce an optional builder for tracers but it would effectively do the same thing).

Author: I rewrote "API and SDK Changes" section to reflect this discussion. Please check.

Author: @yurishkuro please have a look at the updated "API and SDK Changes" and see if you are OK with it.

  // "instrumentation_library_logs" except the logs which have a schema_url
  // specified in the nested InstrumentationLibraryLogs message.
  string schema_url = 3;
}
message InstrumentationLibraryLogs {
  ...
  // This schema_url applies to all logs in the "logs" field regardless of the
  // value of the schema_url field in the outer ResourceLogs message.
  string schema_url = 3;
}
```

The schema_url field in the ResourceSpans, ResourceMetrics, ResourceLogs
messages applies to the contained Resource, Span, SpanEvent, Metric, LogRecord
messages.

The schema_url field in the InstrumentationLibrarySpans message applies to the
contained Span and SpanEvent messages.

The schema_url field in the InstrumentationLibraryMetrics message applies to the
contained Metric messages.

The schema_url field in the InstrumentationLibraryLogs message applies to the
contained LogRecord messages.

If schema_url field is non-empty both in Resource* message and in the contained
InstrumentationLibrary* message then the value in InstrumentationLibrary*
message takes the precedence.

## API and SDK Changes

### Instrumentation Library Schema URL

> OpenTelemetry API need to be changed to allow getting a Tracer/Meter/LogEmitter

Reviewer: ```suggestion
The OpenTelemetry API needs to be changed to allow getting a Tracer/Meter/LogEmitter
```

Author: Fixed.

  they emit. This may be because for example the semantic conventions evolved
  and we want to make our telemetry match the newly introduced conventions.

- In an observability system there may simultaneously exist telemetry sources
  that produce data that conforms to different telemetry schemas because
  different sources evolve at a different pace and are implemented and
  controlled by different entities.

- Telemetry consumers have a need to understand what schema a particular piece
  of received telemetry confirms to. The consumers also need a way to be able to
  interpret the telemetry data that uses different telemetry schemas.

This document proposes a solution to these needs.

## Solution Summary

We believe that the 3 parties described above should be able to evolve
independently over time, while continuously retaining the ability to correctly
work together.

Telemetry Schemas are central to how we make this possible. Here is a summary of
the proposal:

- We introduce a file format for defining telemetry schemas.

- Telemetry schemas are versioned. Over time the schema may evolve and telemetry
  sources may emit data confirming to newer versions of the schema.

> - Telemetry schemas explicitly define transformations that are necessary to

Reviewer 6: +1

  convert telemetry data between different versions of the schema, provided that
  such conversions are possible. When conversions are not possible it
  constitutes a breaking change between versions.

- Telemetry schemas are identified by Schema URLs.

> - Telemetry sources (e.g. instrumentation libraries) should include a schema URL

Reviewer 6: Not explicitly stated (AFAICT);  should the telemetry schema include the version number?

Author: Yes, it is defined later in the doc.

  in the emitted telemetry.

- Telemetry consumers should pay attention to the schema of the received
  telemetry. If necessary, telemetry consumers may transform the telemetry data
  from the received schema version to the target schema version as expected at
  the point of use (e.g. a dashboard may define which schema version it
  expects).

- OpenTelemetry will publish a telemetry schema as part of the specification.
  The schema will contain the list of transformations that semantic conventions

>   undergo. The schema will be available to be referred and downloaded at well

Reviewer 6: Should the "known URL" be the URL of the schema?

Author: Yes, the doc later explains what a Schema URL is in more details.

![Source and Backend](img/0152-source-and-backend.png)

In this example the telemetry source produces spans that comply with version
1.2.0 of OpenTelemetry schema, where "deployment.environment" attribute is used
to record that the span is coming from production.

The telemetry consumer desires to store the telemetry in version 1.1.0 of
OpenTelemetry schema. The schema translator compares the schema_url in the
received span with the desired schema and sees that a version conversion is
needed. It then applies the change that is described in the schema file and
renames the attribute from "deployment.environment" to "environment" before
storing the span.

And here is for example how the schemas can be used to query stored data:

![Query Translate](img/0152-query-translate.png)

### Collector-Assisted Schema Transformation

Here is a somewhat different use case, where the backend is not aware of schemas
and we rely on OpenTelemetry Collector to translate the telemetry to a schema
that the backend expects to receive. The "Schema Translate Processor" is
configured, the target schema_url is specified and all telemetry data that

> passes through the Collector is converted to that target schema:

Reviewer 6: Nit: should this only convert telemetry of a compatible schema (or previous schema version?)

I had thought "schema" would be namespaced/split apart in styles (e.g. HTTP), or is it meant to be ONE schema for ALL semantic conventions?

Author: There is one OpenTelemetry schema for all semantic conventions.

![Collector](img/0152-collector.png)

## Schema URL

Schema URL is an identifier of a Schema. The URL specifies a location of a
[Schema File](#schema-file) that can be retrieved (so it is a URL and not just a
URI) using HTTP or HTTPS protocol.

Fetching the specified URL may return an HTTP redirect status code. The fetcher
MUST follow the HTTP standard and honour the redirect response and fetch the
file from the redirected URL.

> The last part of the URL path is the version number of the schema.

Reviewer 6: Per my comment above, I think this should be in the list of requirements :)

Author: I did not want to overload the Summary section with too many details. Let me see if I can add this.

Author: Reworded the corresponding item in Summary section.

  number.

Correspondingly:

- Consumers of the schema file SHOULD NOT attempt to interpret the schema file
  if the MAJOR version number is different (higher or lower) than what the
  consumer supports.

- Consumers of the schema file SHOULD NOT attempt to interpret the schema file
  if the MINOR version number is higher than what the consumer supports.

- Consumers MAY ignore the PATCH number.

To illustrate this with some examples:

<table>
  <tr>
   <td><strong>File Format Version</strong>
   </td>
   <td><strong>Consumer's Expected Version</strong>
   </td>
   <td><strong>Consumer Can Read?</strong>
   </td>
  </tr>
  <tr>
   <td>1.0.0
   </td>
   <td>1.0.0
   </td>
   <td>yes
   </td>
  </tr>
  <tr>
   <td>1.0.x
   </td>
   <td>1.0.y
   </td>
   <td>yes, for any x and y.
   </td>
  </tr>
  <tr>
   <td>1.a.x
   </td>
   <td>1.b.x
   </td>
   <td>if a&lt;b then yes, otherwise no.
   </td>
  </tr>
  <tr>
   <td>2.0.0
   </td>
   <td>1.x.y
   </td>
   <td>no
   </td>
  </tr>
</table>

## OTLP Changes

To allow carrying the Schema URL in emitted telemetry it is necessary to add a
schema_url field to OTLP messages.

> We add schema_url fields to the following messages:

Reviewer 7: This is an interesting pattern, where we place a new optional field in one of two places, three times over.  Since it's likely to set a precedent, I have a couple of thought experiments: 

1. Instead of adding a new `schema_url` to the _ResourceSignal_ messages here, why not use a resource named `schema.url`? (JSON Schema uses `$schema`, I like that too). This would support schema URLS with no new fields.

2. Instead of adding a new `schema_url` to every of the _InstrumentationLibraryXXX_ messages, could we instead add another place for per-instrumentation-library resources to be expressed?  e.g., `opentelemetry.proto.resource.v1.Resource library_resource = 3;`. Would that be more flexible? I think it might be.

The reason for both of these questions is that I can imagine other kinds of property that are somehow special and if we add a new field for each one, we're in trouble. I'm thinking about information related to the metric collection interval and/or other kinds of processing that may have taken place as telemetry data is collected, information we may want to record either at the global level or at the per-library level.

In transportation systems, the term "Waybill" is used to document what is in transit in a multi-party exchange of goods. We have a similar situation here, where you want to express "what's inside" with a schema, and I strongly suspect there are other kinds of "what's inside" information ahead.

Reviewer 8: I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.

Author: @jmacd 

>Instead of adding a new schema_url to the ResourceSignal messages here, why not use a resource named schema.url? (JSON Schema uses $schema, I like that too). This would support schema URLS with no new fields.

Do you mean a resource attribute named "schema.url"? That would work. The only thing I am not quite sure about is whether it is fair to say that "the telemetry schema is an attribute of a Resource that emits the telemetry".

>Instead of adding a new schema_url to every of the InstrumentationLibraryXXX messages, could we instead add another place for per-instrumentation-library resources to be expressed? e.g., opentelemetry.proto.resource.v1.Resource library_resource = 3;. Would that be more flexible? I think it might be.

Alternatively, we could allow `InstrumentationLibrary` to have arbitrary attributes and record "schema.url" attribute there. I am not quite sure recording a Resource inside an InstrumentationLibraryXXX message reflects our understanding of what a Resource is.

I don't have a strong opinion, open for suggestions.

Author: > I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.

@austinlparker size increase likely is negligible since the schema_url is recorded once per batch of telemetry emitted by instrumentation library.

Reviewer 8: > > I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.
> 
> @austinlparker size increase likely is negligible since the schema_url is recorded once per batch of telemetry emitted by instrumentation library.

Ah, ok, it read like you could apply multiple `schema.url` attributes. Looking at OTLP, it seems like adding this field to `InstrumentationLibrary` would make the most sense.

Author: >Maybe we just handle them as attributes rather than as resources and build semantic conventions up for them.

Yes, this is a possible option.

Author: I have put some more thought into this. While I believe it is a possible option I think it a tiny bit less desirable than using a distinct protocol field in OTLP. If the schema url is recorded simply as an attribute we need to come up with a name for the that attribute, let's assume for example it is `schema.url`. But what if we want to change the name of this attribute? We can't, this will be the one attribute which is not described and is not handled by the schema. This becomes an exception. I do not quite like having exceptions like this.
I think having a distinct new field in OTLP that is part of the proto is warranted in this case since it is a meta information that describes the attributes, it is not the same kind of information as all other attributes of a resource are.
So, for this reason my slight preference is to keep my original proposal.
To be clear: this does not prevent us from introducing attributes for instrumentation libraries. But that's a separate topic and can be done in a separate OTEP unrelated to this one.

Reviewer 9: I agree.

Reviewer 8: I concur with your logic (the "meta information that describes attributes" part). Perhaps you could add a note or errata that indicates why this information is special enough to warrant a new distinct field, in an attempt to keep this decision from establishing precedent? I think I'm more concerned about the potential for this decision to create a ripple effect, where we're introducing many versions of OTLP and causing frission for implementors and end-users alike.

Author: OK, Let me see if I can add something to the document.

To be clear: introducing `schema_url` as a new field in OTLP is a fully backwards compatible change. The default value for `schema_url` will be the empty string (as it normally works for new fields in Protobufs) which works great for our purposes. Empty `schema_url` indicates that the schema is unspecified (as you would expect from any pre-existing software that doesn't know about the concept of schema).
This is one of the fortunate cases when OTLP sender implementations and OTLP receiver implementations don't need to coordinate anything and can start supporting `schema_url` at their leisure. :-)

depend on the composition of the data, the volume of transformations done during
conversion, programming language used, etc. However, we feel that this one data
point is still useful for understanding that the potential impact is acceptable
(and likely will be a very small percentage of total processing most telemetry
consumers do).

## Open Questions

- Do we need to support a concept of "pre-release" semantic conventions that can
  be broken freely and are not part of strict schema checks? Supposedly, this is
  already possible simply by avoiding introducing a new schema version when such
  semantic conventions change.

- How to make it easy for instrumentation libraries to refer to a particular
  OpenTelemetry schema version and also make sure any semantic convention
  helpers the library uses (e.g. constants that define the semantic conventions)
  match exactly that same schema version? One possible solution is to introduce
  helper packages per schema version that the libraries can use, e.g. constants
  that define the semantic conventions and the corresponding schema version url.
  This should be likely the topic for a follow-up OTEP.

> - Should we make it possible to include the entire Schema File in OTLP requests

Reviewer 7: [Thinking about how JSON Schema encourages re-use with a `$ref` keyword], I could imagine a "Yes" answer here that does not require any more protocol change than what you have here.

We could specify to use a schema URL like `resource://some.identifier` and then use the `some.identifier` resource value as a schema--that might work except for resource value size restrictions.

In a logging system, (maybe, if it were up to me, ...), I might suggest the process log its schema early on in the process lifetime and then use a schema URL that refers backwards in the log.

Author: I like this idea.

Author: Taking a note for the future. Resolving for now, since it is not mandatory for this initial proposal.

The suggestion was to record the encoding of the Body field of Log Records,
which can be also implemented as a new setting in the "current state" of the
"logs" section in the schema file.

### Other Transformation Types

This OTEP introduces a limited set of transformations, while deliberately
keeping the type of transformations to a minimum. It is easy to see that
telemetry can evolve in other, more sophisticated ways over time (see e.g.
[this list](https://github.com/open-telemetry/opentelemetry-specification/issues/1324)).

More transformation types can be added in the future when there is a need to
represent a particular change in the telemetry schema that is not possible with
the current set of transformations. However, care should be taken to avoid
introducing unnecessary or overly complicated transformation types, especially
ones that cannot be applied on a local portion of telemetry data and requires a
full state (e.g. transformations, such as aggregations, etc), since such
transformations can place significant implementation burden on telemetry sources
that wish to support the notion of telemetry schemas.

> ## Alternates Considered

Reviewer 10: This is a rather complex and powerful approach. To justify it, I think there is one more alternative to consider (with corresponding subsection here): Have a strict policy to never rename an attribute once it becomes stable. While this may lead to sub-optimal names, it would be much simpler to implement and have not negative performance impact.

Author: This is surely up for debate.
I think I have heard from enough sources that fossilizing semantic conventions and emitted telemetry schema is not an option. IMO, long term it is an impossible position. People make mistakes and want to be able to fix them. The eventual surface area of semantic convention and telemetry schemas as a whole is much bigger than the API.
Forcedly fixing the API is achievable (especially given that we have tools like compilers that help enforce it). Fixing the semantic convention I think is much more difficult (especially given that there is no fully machine-readable definition of Otel schema and there are no tools that can help enforce that schema). We could introduce such machine-readable schema and tools, but the effort for Otel will be bigger than what this proposal asks for and still won't be as useful for enabling the evolution of schemas.

Reviewer 10: That sounds reasonable, but I would add a subsection on this to the "Alternates Considered".

Author: Sounds good, I will add.

Author: Added a subsection "Freeze Schema" in "Alternates Considered".

  cannot be handled by aliases).

- Over time more and more data should be published increasing the volume of the
  published data. Depending on how exactly the aliases handled we may need to
  duplicate a lot (e.g. if metric name has change we may have to produce the
  entire same metric twice, essentially duplicating the traffic) or may require
  breaking changes in the OpenTelemetry API or protocol to allow aliases
  natively.

### Schema Negotiation

Instead of performing transformations of the schema let telemetry consumers
(backends) negotiate the version they expect to receive from the telemetry
source.

The benefit of this approach is that the backend can support just one schema,
there is no need to define transformation rules or do the transformations at
all.

The downsides are:

- Requires telemetry sources to be able to emit telemetry in multiple different
  schema versions, which is a significant burden on the telemetry sources. This
  is likely a dealbreaker.

- Requires a communication channel from telemetry consumers to telemetry sources
  which currently does not exist (all communication is currently
  one-directional, from sources to consumers).

> ## Prior Art

Reviewer 7: Fantastic. 😀

- Telemetry sources over time may want to change the schema of the telemetry
  they emit. This may be because for example the semantic conventions evolved
  and we want to make our telemetry match the newly introduced conventions.

- In an observability system there may simultaneously exist telemetry sources
  that produce data that conforms to different telemetry schemas because
  different sources evolve at a different pace and are implemented and
  controlled by different entities.

- Telemetry consumers have a need to understand what schema a particular piece
  of received telemetry confirms to. The consumers also need a way to be able to
  interpret the telemetry data that uses different telemetry schemas.

This document proposes a solution to these needs.

## Solution Summary

We believe that the 3 parties described above should be able to evolve
independently over time, while continuously retaining the ability to correctly
work together.

Telemetry Schemas are central to how we make this possible. Here is a summary of
the proposal:

- We introduce a file format for defining telemetry schemas.

- Telemetry schemas are versioned. Over time the schema may evolve and telemetry

>   sources may emit data confirming to newer versions of the schema.

Reviewer 2: Should Telemetry Schema versions correspond to versions of the OpenTelemetry spec? Or would those be two different (but linked) lines of versions?

Author: There is no direct match between Otel spec version numbers and Otel schema version numbers. There is no need to have such coupling, the two should be able evolve independently. 
Yes, a particular Otel spec X version will naturally correspond a particular Otel schema version Y (since semantic conventions are part of the spec and schemas are tied to semantic conventions), however X may and likely will be different from Y.

Reviewer 5: This feels rather confusing to me. What's the argument against keeping schema and spec version number the same?

Author: If we want the version numbers the same then we have to align releases of the spec with changes in schema. I was thinking that keeping these decoupled gives us more freedom to evolve the 2 independently.

Not a strong opinion though. I think using the same version numbers would work too. It just means that in some cases we will release a new schema version when there are no changes to it (because it is time to release a new spec version but nothing is changed in semantic convention). That's fine, we just need to be careful not to forget to keep schema version number in sync with spec version number (an automated pre-release check would be preferable).
Likewise, sometimes there is probably nothing new in the spec other than changes to semantic conventions, so we will have to release a new spec version to get a new version number we can use for new schema.

If we feel keeping version numbers in sync is less confusing, then I don't mind it, that can work.

Reviewer: I think requiring a spec release for an updated schema would be a very good thing.

Reviewer 2: I'd be in favor of sharing the same version numbers. I expect people writing instrumentation according to a certain version of the specification. It seems natural to me that this version is the same as the schema version.

Author: The version number recommendations section I have in the doc will no longer work in that case. Let me think about this.

Reviewer 11: Sharing them means we will have as many schema versions as API ones, which feels odd and I don't remember seeing a schema with that many versions.

Author: I spent a bit time on this and I believe I agree that it is preferable to keep the schema version numbers in sync with the spec version numbers.

@carlosalberto I understand your concern. In some cases there will be no actual changes in the schema, but we will still bump the version number to keep it in sync with spec version number. It will mean adding one line to the schema YAML file. We can also automatically verify that this line is correctly added, so I think it is not a big problem and the trade-off is worth it to avoid confusion and the need to explain mismatching version numbers.
I will modify the proposal to reflect this.

Author: I changes sections "Schema Version Number" and "OpenTelemetry Schema", please have a look.

Reviewer 2: Thanks. I think it will make life easier for people writing instrumentation: they need to keep track of only two versions (OpenTelemetry spec and language specific API) instead of three (spec, language API, and telemetry schema).

Author: @yurishkuro @carlosalberto are you OK with the changes in "Schema Version Number" and "OpenTelemetry Schema" sections?

Reviewer 11: @tigrannajaryan I'm fine with it. We will have to validate it once this hits the Specification + implementation, so we can still revert if truly needed.

overrides the application-wide schema url as described [here](#otlp-changes).

## OpenTelemetry Schema 1.0.0

The following is the full content of Schema File for OpenTelemetry Schema 1.0.0:

```yaml
file_format: 1.0.0
schema_url: https://opentelemetry.io/schemas/1.0.0
versions:
  1.0.0:
```

Since this is the first version of OpenTelemetry schema there are no changes
since the previous version (because there is no previous version) so the section
1.0.0 is empty.

This file SHOULD be available for retrieval at
[https://opentelemetry.io/schemas/1.0.0](https://opentelemetry.io/schemas/1.0.0)

All OpenTelemetry instrumentation solutions will follow this schema.

Changes to OpenTelemetry semantic conventions SHOULD be handled as described in
the [Schema Version Number](#schema-version-number) section. For example,
addition of a completely new optional attribute to the semantic conventions
SHOULD NOT require a new telemetry schema version. Similarly, a removal of an
optional attribute SHOULD NOT require a new schema version, removal or addition
of a required attribute to an existing entity SHOULD require a new schema

> version. Generally, any changes to the semantic conventions schema that do not

Reviewer 3: If we add a new required attribute to the semantic convention, how will a new schema version file look like? It still will be empty, right? Because there is no transformation that changes previous version into a new one?

Author: Correct. The schema "changes" will be empty. However, I believe it is still valuable to record it as a new schema version, simply to indicate that it is different. The telemetry consumers are free to use or ignore that fact.
In the future we may add ability to explicitly record in the "changes" section the fact that a required attribute was added.

which can be also implemented as a new setting in the "current state" of the
"logs" section in the schema file.

### Other Transformation Types

This OTEP introduces a limited set of transformations, while deliberately
keeping the type of transformations to a minimum. It is easy to see that
telemetry can evolve in other, more sophisticated ways over time (see e.g.
[this list](https://github.com/open-telemetry/opentelemetry-specification/issues/1324)).

More transformation types can be added in the future when there is a need to
represent a particular change in the telemetry schema that is not possible with
the current set of transformations. However, care should be taken to avoid
introducing unnecessary or overly complicated transformation types, especially
ones that cannot be applied on a local portion of telemetry data and requires a
full state (e.g. transformations, such as aggregations, etc), since such
transformations can place significant implementation burden on telemetry sources
that wish to support the notion of telemetry schemas.

## Alternates Considered

### Freeze Schema

Instead of introducing formal schemas, schema files and version we can require

> that the instrumentation once it is created never changes. Attributes are never

Reviewer 10: As long as the instrumentation conforms to the semantic conventions before and after, that should be fine, shouldn't it?

I think rather than freezing the schema completely, we only need to ban changes that would create an entry in the YAML file defined in this OTEP. So renaming is not allowed, adding new attributes is. Changing the meaning of an attribute is also not allowed, but I think this can also not automatically handled with the schema YAML, although the version specified there could then be used in the backend for version-specific handling.

Author: Yes, certain changes would be fine with the approach you propose. I still don't think this alternate approach is sufficient. I don't believe it solves the problems we want to solve. In my opinion being able to rename attributes or metrics is a basic requirement. Given how much time we spend arguing about attribute names here in Otel semantic conventions and how much uncertainty is there that the name is right even after the attribute is accepted I believe we will inevitably run into a situation when name changes are wanted.

I saw schema troubles in the past. For example with metrics emitted by Otel Collector, we renamed metrics and broke Collector dashboards in a proprietary monitoring tool. When schemas are not explicitly declared and are not a visible concept they still change and things break, just less explicitly visibly so.

Again, we can argue that we should be careful and come up with semantic conventions that we think are right and then never change them. I think it is an attempt to achieve perfection that is very difficult to do and at the same time increases the barrier to introducing semantic conventions because the cost of making a mistake is very high. I think we should lower that cost a bit. It should still be expensive to change semantic conventions (we don't want to encourage that too much), but it should not be impossible.

conventions) match exactly that same schema version? One possible solution is to
introduce helper packages per schema version that the libraries can use, e.g.
constants that define the semantic conventions and the corresponding schema
version url. This should be likely the topic for a follow-up OTEP.

### Application-wide Schema URL

The SDK interface MUST provide a way for the user to optionally set an
application-wide Schema URL. This Schema URL will be populated in all
ResourceSpans, ResourceMetrics and ResourceLogs messages emitted by OTLP
Exporter.

If the user does not set an application-wide Schema URL then the current Schema
URL MUST be populated by OTLP Exporter in the messages, where "current" means
the version of OpenTelemetry Schema against which the SDK is coded.

Note that if there is a schema url associated with instrumentation library it
overrides the application-wide schema url as described [here](#otlp-changes).

## OpenTelemetry Schema

OpenTelemetry publishes it own schema at
`https://opentelemetry.io/schemas/<version>`. The version number of the schema
is the same as the specification version number which publishes the schema.

> Every time a new specification version is released a corresponding new schema

Reviewer 5: It may be useful to require even stronger coupling. Since both schemas and semantic conventions are machine-readable, there could be a simple validation script that makes sure that the new version of the schema accurately reflects the changes to the spec, e.g. if an attribute was removed / renamed from the spec, something about that has to be present in the schema and we should block the release otherwise.

Author: Yes, validation is a good idea.

              # in the previous version, the values are the new attribute name
              # starting from this version.
              peer.service: peer.service.name

    span_events:
      # Definitions that apply to Span Event data type.
      changes:
        - rename_events:
            # The keys are old event name used in the previous version, the
            # values are the new event name starting from this version.
            name_map: {stacktrace: stack_trace}

        - rename_attributes:
            attribute_map:
              peer.service: peer.service.name
            apply_to_events:
              # Optional event names to apply to. If empty applies to all events.
              - exception.stack_trace

    metrics:
      # Definitions that apply to Metric data type.
      changes:
        - rename_metrics:
            # map of key/values. The keys are the old metric name used
            # in the previous version, the values are the new metric name
            # starting from this version.
            container.cpu.usage.total: cpu.usage.total
            container.memory.usage.max: memory.usage.max

>         - rename_labels:

Author: Thanks. I'll wait until the spec is updated and will update here as well.
